extensions/amp-stream-gallery/1.0/stream-gallery.js

```diff
@@ -0,0 +1,216 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {BaseCarousel} from '../../amp-base-carousel/1.0/base-carousel';
+import {
+  useCallback,
+  useLayoutEffect,
+  useMemo,
+  useRef,
+  useState,
+} from '../../../src/preact';
+import {useStyles} from './stream-gallery.jss';
+
+const DEFAULT_MEASUREMENT = {
+  visibleCount: 1,
+  maxContainerWidth: Number.MAX_VALUE,
+};
+
+/**
+ * @param {!StreamGalleryDef.Props} props
+ * @return {PreactDef.Renderable}
+ */
+export function StreamGallery({
+  arrowPrev: customArrowPrev,
+  arrowNext: customArrowNext,
+  children,
+  extraSpace,
+  insetArrowVisibility,
+  loop,
+  maxItemWidth = Number.MAX_VALUE,
+  minItemWidth = 1,
+  maxVisibleCount = Number.MAX_VALUE,
+  minVisibleCount = 1,
+  outsetArrows,
+  peek = 0,
+  snap,
+  style,
+  ...rest
+}) {
+  const ref = useRef(null);
+  const [measurements, setMeasurements] = useState(DEFAULT_MEASUREMENT);
+  const arrowPrev = useMemo(
+    () =>
+      customArrowPrev ?? <DefaultArrow by={-1} outsetArrows={outsetArrows} />,
+    [customArrowPrev, outsetArrows]
+  );
+  const arrowNext = useMemo(
+    () =>
+      customArrowNext ?? <DefaultArrow by={1} outsetArrows={outsetArrows} />,
+    [customArrowNext, outsetArrows]
+  );
+
+  const measure = useCallback(
+    (containerWidth) =>
+      getVisibleCount(
+        maxItemWidth,
+        minItemWidth,
+        maxVisibleCount,
+        minVisibleCount,
+        children.length,
+        peek,
+        containerWidth
+      ),
+    [
+      maxItemWidth,
+      minItemWidth,
+      maxVisibleCount,
+      minVisibleCount,
+      children.length,
+      peek,
+    ]
+  );
+
+  // Adjust visible slide count when container size or parameters change.
+  useLayoutEffect(() => {
+    const {node} = ref.current;
+    if (!node) {
+      return;
+    }
+    const observer = new ResizeObserver((entries) => {
+      const last = entries[entries.length - 1];
+      setMeasurements(measure(last.contentRect.width));
+    });
+    observer.observe(node);
+    return () => observer.disconnect();
+  }, [measure]);
+
+  return (
+    <BaseCarousel
+      advanceCount={Math.floor(measurements.visibleCount)}
+      arrowPrev={arrowPrev}
+      arrowNext={arrowNext}
+      controls={insetArrowVisibility}
+      loop={loop}
+      outsetArrows={outsetArrows}
+      snap={snap}
+      ref={ref}
+      style={{
+        ...style,
+        flexGrow: 1,
+        maxWidth:
+          measurements.maxContainerWidth >= Number.MAX_VALUE
+            ? ''
+            : measurements.maxContainerWidth,
+        justifyContent: extraSpace === 'around' ? 'center' : 'initial',
```

Comment:
This might be an important issue. It's hard to use flexbox in styles because the vendor-prefixing is very complicated. Is there a way to push `flexGrow`, `justifyContent`, and similar to JSS? We can always allow a mode on `BaseCarousel` and let it apply the stylesheet class.

Reply:
Moved these to JSS, PTAL.